File: packages/reference/android/src/test/java/github/penguin/reference/TestClasses.java
package github.penguin.reference;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import github.penguin.reference.async.Completable;
import github.penguin.reference.async.Completer;
import github.penguin.reference.reference.InstanceManager;
import github.penguin.reference.reference.PairedInstance;
import github.penguin.reference.reference.TypeChannelHandler;
import github.penguin.reference.reference.TypeChannelMessageDispatcher;
import github.penguin.reference.reference.TypeChannelMessenger;

public class TestClasses {
  public static class TestMessenger extends TypeChannelMessenger {
    public final TestMessageDispatcher testMessageDispatcher = new TestMessageDispatcher();
    public final TestInstanceManager testInstancePairManager = new TestInstanceManager();
    public final TestHandler testHandler = new TestHandler();

    public TestMessenger() {
      registerHandler("test_channel", testHandler);
    }

    @Override
    public TypeChannelMessageDispatcher getMessageDispatcher() {
      return testMessageDispatcher;
    }

    @NonNull
    @Override
    public InstanceManager getInstanceManager() {
      return testInstancePairManager;
    }
  }

  public static class TestMessageDispatcher implements TypeChannelMessageDispatcher {
    @Override
    public Completable<Void> sendCreateNewInstancePair(String channelName, PairedInstance pairedInstance, List<Object> arguments, boolean owner) {
      return new Completer<Void>().complete(null).completable;
    }

    @Override
    public Completable<Object> sendInvokeStaticMethod(String channelName, String methodName, List<Object> arguments) {
      return new Completer<>().complete("return_value").completable;
    }

    @Override
    public Completable<Object> sendInvokeMethod(String channelName, PairedInstance pairedInstance, String methodName, List<Object> arguments) {
      return new Completer<>().complete("return_value").completable;
    }

    @Override
    public Completable<Void> sendDisposeInstancePair(PairedInstance pairedInstance) {
      return new Completer<Void>().complete(null).completable;
    }
  }

  public static class TestHandler implements TypeChannelHandler<TestClass> {
    public final TestClass testClassInstance = new TestClass();

    @Override
    public List<Object> getCreationArguments(TypeChannelMessenger manager, TestClass instance) {
      return Collections.emptyList();
    }

    @Override
    public TestClass createInstance(TypeChannelMessenger manager, List<Object> arguments) {
      return testClassInstance;
    }

    @Override
    public Object invokeStaticMethod(TypeChannelMessenger manager, String methodName, List<Object> arguments) {
      return "return_value";
    }

    @Override
    public Object invokeMethod(TypeChannelMessenger manager, TestClass instance, String methodName, List<Object> arguments) {
      return "return_value";
    }
  }

  public static class TestInstanceManager extends InstanceManager {
    @Override
    protected String generateUniqueInstanceId(Object instance) {
      return "test_instance_id";
    }
  }

  public static class TestClass {
  }

  public static class TestListener<T> implements Completable.OnCompleteListener<T> {
    public T result;

    @Override
    public void onComplete(T result) {
      this.result = result;
    }

    @Override
    public void onError(Throwable throwable) {
    }
  }
}

File: packages/reference/test/test_classes.dart
import 'package:reference/reference.dart';

class TestMessenger extends TypeChannelMessenger {
  TestMessenger() {
    registerHandler('test_channel', testHandler);
  }

  late final TestHandler testHandler = TestHandler();

  @override
  final TestMessageDispatcher messageDispatcher = TestMessageDispatcher();

  @override
  final TestInstanceManager instanceManager = TestInstanceManager();
}

class TestHandler with TypeChannelHandler<TestClass> {
  final TestClass testClassInstance = TestClass();

  @override
  TestClass createInstance(
    TypeChannelMessenger manager,
    List<Object?> arguments,
  ) {
    return testClassInstance;
  }

  @override
  Object? invokeMethod(
    TypeChannelMessenger manager,
    TestClass instance,
    String methodName,
    List<Object?> arguments,
  ) {
    return 'return_value';
  }

  @override
  Object? invokeStaticMethod(
    TypeChannelMessenger manager,
    String methodName,
    List<Object?> arguments,
  ) {
    return 'return_value';
  }
}

class TestMessageDispatcher with TypeChannelMessageDispatcher {
  @override
  Future<void> sendCreateNewInstancePair(
    String handlerChannel,
    PairedInstance remoteReference,
    List<Object?> arguments, {
    required bool owner,
  }) {
    return Future<void>.value();
  }

  @override
  Future<Object?> sendInvokeMethod(
    String handlerChannel,
    PairedInstance remoteReference,
    String methodName,
    List<Object?> arguments,
  ) {
    return Future<String>.value('return_value');
  }

  @override
  Future<Object?> sendInvokeStaticMethod(
    String handlerChannel,
    String methodName,
    List<Object?> arguments,
  ) {
    return Future<String>.value('return_value');
  }

  @override
  Future<void> sendDisposeInstancePair(PairedInstance pairedInstance) {
    return Future<void>.value();
  }
}

class TestClass {}

class TestInstanceManager implements InstanceManager {
  final Map<Object, String> instanceToInstanceId = <Object, String>{};
  final Map<String, Object> instanceIdToInstance = <String, Object>{};

  @override
  Object? getInstance(String instanceId) {
    return instanceIdToInstance[instanceId];
  }

  @override
  String? getInstanceId(Object instance) {
    return instanceToInstanceId[instance];
  }

  @override
  bool containsInstance(Object instance) {
    return instanceToInstanceId.containsKey(instance);
  }

  @override
  void removeInstance(String instanceId) {
    final Object? instance = instanceIdToInstance.remove(instanceId);
    instanceToInstanceId.remove(instance);
  }

  @override
  bool addStrongReference({required Object instance, String? instanceId}) {
    if (containsInstance(true)) return false;

    final String newId = instanceId ?? generateUniqueInstanceId(instance);
    instanceToInstanceId[instance] = newId;
    instanceIdToInstance[newId] = instance;
    return true;
  }

  @override
  bool addWeakReference({
    required Object instance,
    String? instanceId,
    required void Function(String instanceId) onFinalize,
  }) {
    if (containsInstance(true)) return false;
    final String newId = instanceId ?? generateUniqueInstanceId(instance);
    return addStrongReference(instance: instance, instanceId: newId);
  }

  @override
  String generateUniqueInstanceId(Object instance) {
    return 'test_reference_id';
  }
}
